File: MLE_test.m
clear; clc; close all;
a = -1;b=-1;c=30;
f = @(x,y) a*x+b*y+c;
x_t = 1:10;
y_t = 1:10;
fall_point = [12,12,0];
z_t = f(x_t,y_t);
X = [x_t',y_t',z_t'];
X = [X;fall_point];
N = ndims(X);

X_ave = mean(X,1);
dx = bsxfun(@minus,X,X_ave);
C = (dx' * dx)/(N-1);
[R,D] = svd(C,0);
D = diag(D);
R2 = D(1)/sum(D);

x = dx * R(:,1);
x_min = min(x);
x_max = max(x);
difx = x_max - x_min;
Xa = (x_min - 0.05 * difx) * R(:,1)' + X_ave;
Xb = (x_max + 0.05 * difx) * R(:,1)' + X_ave;

end_x = [Xa;Xb];

figure('color','w')
axis equal
hold on;
plot3(end_x(:,1),end_x(:,2),end_x(:,3),'-r');
plot3(X(:,1),X(:,2),X(:,3),'-g');

% coefficient = polyfit([x_t,fall_point(1);y_t,fall_point(2)],[[z_t,0]],1)
% f_fit = @(x,y) coefficient(1)*x + coefficient(2)*y + coefficient(3);
% 
% x_y = [1:15;1:15];
% figure(1);
% hold on;
% plot(x_y,f(x_y(1,:),x_y(2,:)),'-g');
% plot(x_y,f_fit(x_y(1,:),x_y(2,:)),'-r');

File: MLE_test.py
import numpy as np
import matplotlib.pyplot as plt

if __name__ == "__main__":
    a , b, c = -1, -1, 30
    x_t = np.arange(10).reshape(1,10)
    y_t = np.arange(10).reshape(1,10)
    z_t = a* x_t + b * y_t + c
    fall_point = np.array([12,12,0])
    X = np.column_stack([x_t.transpose(),y_t.transpose(),z_t.transpose()])#.reshape(10,3)
    X = np.row_stack([X,fall_point])
    N = X.shape[0]

    X_ave = np.mean(X,0)
    dx = X - X_ave.reshape(1,3)
    w = np.array([1,1,1,1,1,1,1,1,1,1,10]).reshape(11,1)
    dx = np.multiply(w,dx)
    C = dx.transpose().dot(dx) /(N-1)
    U,S,V = np.linalg.svd(C)
    x = dx.dot(U[:,0])
    x_min = np.min(x)
    x_max = np.max(x)
    difx = x_max - x_min
    Xa = (x_max - 0.19*difx) * U[:,0].transpose() + X_ave
    Xb = (x_max + 0.05 * difx) * U[:,0].transpose() + X_ave
    end_x = np.row_stack([Xa,Xb])

    #print(X)
    fig = plt.figure()
    ax = fig.gca(projection='3d')
    ax.scatter(X[:,0],X[:,1],X[:,2])
    ax.plot(end_x[:,0],end_x[:,1],end_x[:,2],'r')
    plt.show()

    